CMSgov/easi-app
# [EASI-3047] Generate presigned URLs for TRB attached docs on-demand

# EASI-3047

## Changes and Description

- Instead of getting the presigned URL from the backend when the page with a `<DocumentsTable>` is loaded (which runs into issues if the user waits on the page long enough for the presigned URL to expire), have the "View" button query the backend for the URL when clicked, then immediately download the file from it.


## How to test this change

### Testing basic functionality

1. Create a TRB request with an attached document.
2. Run `scripts/dev minio:clean` to mark the attachment as clean (virus-free), making it available to view/download.
3. In the admin view for that request, go to the "Supporting documents" sub-page.
4. Click "View" for the document - you should be able to download the attached document.

### Verifying that this avoids issues with the presigned URL expiring

1. Change the expiration time for presigned URLs - in `pkg/upload/upload.go`, in `NewGetPresignedURL()`, change the call `req.Presign(15 * time.Minute)` to `req.Presign(1 * time.Minute)` to make presigned URLs expire after 1 minute.
2. As above, create a request, attach a document, and mark it clean.
3. Go to the "Supporting documents" sub-page.
4. Wait for more than 1 minute, then click "View" - you should still be able to download the document.

## Notes
- There's no loading indicator when fetching a presigned URL; the user clicks "View", the UI doesn't change at all or show any kind of indicator while it's fetching the URL. I'm not sure if this'll be a potential issue in deployed environments.
- The error handling if the query to fetch a presigned URL fails is a little unpolished; currently, the `showMessage()`-based approach only displays an error in admin views, not requester views.
- The additions of `MessageProvider` in test files seemed necessary to fix test failures [like in this CI run](https://github.com/CMSgov/easi-app/actions/runs/5478980172/jobs/9980262490). I based the addition off of other TRB unit tests.
